procrustes/kopt.py

```diff
@@ -34,61 +34,55 @@
-    ref_error : float
-        The reference error value.
-    perm : ndarray, optional
-        The permutation array which remains to be processed with k-opt local search. Default is the
-        identity matrix with the same shape of array_a.
-    kopt_k : int, optional
-        Defines the oder of k-opt heuristic local search. For example, kopt_k=3 leads to a local
-        search of 3 items and kopt_k=2 only searches for two items locally. Default=3.
-    kopt_tol : float, optional
-        Tolerance value to check if k-opt heuristic converges. Default=1.e-8.
+    a : ndarray
+        The 2D-array :math:`\mathbf{A}` which is going to be transformed.
+    b : ndarray
+        The 2D-array :math:`\mathbf{B}` representing the reference matrix.
+    p : ndarray, optional
+        The 2D-array :math:`\mathbf{P}` representing the initial permutation matrix. If `None`, the
+        identity matrix is used.
+    k : int, optional
+        The order of the permutation. For example, `k=3` swaps all possible 3-permutations of the
+        given p matrix.
+    tol : float, optional
+        Convergence threshold of the heuristic algorithm.
 
     Returns
     -------
     perm : ndarray
         The permutation array after optimal heuristic search.
-    kopt_error : float
+    error : float
         The error distance of two arrays with the updated permutation array.
     """
-    if kopt_k < 2:
-        raise ValueError("Kopt_k value must be a integer greater than 2.")
-    # if perm is not specified, use the identity matrix as default
-    if perm is None:
-        perm = np.identity(np.shape(array_a)[0])
-    num_row = perm.shape[0]
-    kopt_error = ref_error
+    if k < 2 or not isinstance(k, int):
+        raise ValueError(f"Argument k must be a integer greater than 2. Given k={k}")
+    # assign p to be an identity array, if not specified
+    if p is None:
+        p = np.identity(np.shape(a)[0])
+    num_row = p.shape[0]
+    error = compute_error(a, b, p, p)
     # all the possible row-wise permutations
-    for comb in it.combinations(np.arange(num_row), r=kopt_k):
-        for comb_perm in it.permutations(comb, r=kopt_k):
+    for comb in it.combinations(np.arange(num_row), r=k):
+        for comb_perm in it.permutations(comb, r=k):
             if comb_perm != comb:
-                perm_kopt = deepcopy(perm)
-                perm_kopt[comb, :] = perm_kopt[comb_perm, :]
-                e_kopt_new = compute_error(array_a, array_b, perm_kopt, perm_kopt)
-                if e_kopt_new < kopt_error:
-                    perm = perm_kopt
-                    kopt_error = e_kopt_new
-                    if kopt_error <= kopt_tol:
+                p_new = deepcopy(p)
+                p_new[comb, :] = p_new[comb_perm, :]
+                error_new = compute_error(a, b, p_new, p_new)
+                if error_new < error:
+                    p, error = p_new, error_new
+                    if error <= tol:
                         break
```

Comment:
1. Upon finding a better permutation matrix, shouldn't we start the k-opt from the beginning? The current code doesn't start the `comb` and `comb_perm` swaps for the new `p` matrix.
2. Isn't this a brute force algorithm? Then, I don't understand the goal of having a `tol` argument. Based on my understanding, we should just try **all** k-permutations.

Reply:
1. For the previous implementation version, we didn't run the k-opt from the beginning again once we find an improvement. But I think about it again and it seems we should.
2. This is a greedy algorithm. Using `tol` was an attempt to set up an early stop as `k-opt` is computationally expensive. We set `tol` to be very small (`1.e-8`) and in most cases, it won't affect anything.

Reply:
1. Fixed in b040f06
2. `tol` was removed as a user-defined argument in 19c4534, but it kept in the algorithm to stop the search when a perfect permutation matrix is found.